jlm/rvsdg/NodeNormalization.hpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_NODENORMALIZATION_HPP
+#define JLM_RVSDG_NODENORMALIZATION_HPP
+
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+class output;
+
+/**
+ * \brief Normalizes the operands of an operation.
+ *
+ * The usage expectation is that IsApplicable() is invoked to determine whether operands can be
+ * normalized BEFORE ApplyNormalization() is invoked.
+ *
+ * @tparam TOperation The operation for which to normalize the operands.
+ */
+template<class TOperation>
+class NodeNormalization
+{
+  static_assert(
+      std::is_base_of<Operation, TOperation>::value,
+      "Template parameter TOperation must be derived from jlm::rvsdg::operation.");
+
+public:
+  virtual ~NodeNormalization() noexcept = default;
+
+  NodeNormalization() = default;
+
+  NodeNormalization(const NodeNormalization &) = delete;
+
+  NodeNormalization(NodeNormalization &&) = delete;
+
+  NodeNormalization &
+  operator=(const NodeNormalization &) = delete;
+
+  NodeNormalization &
+  operator=(NodeNormalization &&) = delete;
+
+  /**
+   * Determines whether the operands can be normalized.
+   *
+   * \note While the method can modify internal state, it is expected to be re-entrant.
+   *
+   * @param operation The operation for which to normalize the operands.
+   * @param operands The operands corresponding to the operation.
+   *
+   * @return True, if the operands can be normalized, otherwise false.
+   */
+  virtual bool
+  IsApplicable(const TOperation & operation, const std::vector<output *> & operands) = 0;
+
+  /**
+   * \brief Performs the normalization of the operands.
+   *
+   * @param operation The operation for which to normalize the operands.
+   * @param operands The operands corresponding to the operation.
+   *
+   * @return The normalized operands.
+   */
+  virtual std::vector<output *>
+  ApplyNormalization(const TOperation & operation, const std::vector<output *> & operands) = 0;
```

Comment:
Why not combine IsApplicable and ApplyNormalization into single method that returns std::optional<std::vector<output *>> ?

Looking at the single implementation that you wrote already it appears they are going to have a lot of code redundancies anyways.

Reply:
I was considering this interface originally as well, but decided against it for now. It would have meant that I would not just have to decouple the normalizations from the normal form, but also extensively edit them to fit this new interface, potentially leading to bugs etc. As it is right now, it is more or less a copy-paste exercise to copy the code over from the old methods in the new interface. Once, the interface is up and we have thrown the normalization form, I am planning to go towards this interface. It would also be able to more easily go there, as I could convert each normalization individually at the time.

TLDR: It believed it to be simpler and safer to go this route first, before going to this other interface.

Reply:
well if you have a function returning `std::optional<...>` then it can serve both purposes -- call sites that want to check first simply call it twice, so you have an immediate transition path

I don't understand the "convert each normalization individually" bit -- since you did not define a common underlying interface for all of them, they are entirely decoupled anyways. The "base class" you defined is templated, so cannot be a common interface.

Reply:
My main goal is to get the old node normalization interface out. I did not want to refactor the individual normalizations on top of that.